emission/analysis/userinput/matcher.py

```diff
@@ -63,32 +63,88 @@ def handle_single_most_recent_match(confirmed_trip, ui):
-    if "trip_addition" not in confirmed_trip["data"] or \
-        confirmed_trip["data"]["trip_addition"] is None:
-        confirmed_trip["data"]["trip_addition"] = []
+    if "additions" not in confirmed_trip["data"] or \
+        confirmed_trip["data"]["additions"] is None:
+        confirmed_trip["data"]["additions"] = []
     if "status" not in ui.data or ui.data.status == ecwtui.InputStatus.ACTIVE:
-        confirmed_trip["data"]["trip_addition"].append(ui)
+        confirmed_trip["data"]["additions"].append(ui)
     elif ui.data.status == ecwtui.InputStatus.DELETED:
-        after_del_list = [ta for ta in confirmed_trip["data"]["trip_addition"] if ta["match_id"] != ui["match_id"]]
-        confirmed_trip["data"]["trip_addition"] = after_del_list
+        after_del_list = [ta for ta in confirmed_trip["data"]["additions"] if ta["match_id"] != ui["match_id"]]
+        confirmed_trip["data"]["additions"] = after_del_list
     else:
         # TODO: Decide whether to error or to warn here
         logging.warn("Invalid status found in user input %s, moving forward anyway" % ui)
 
+def create_composite_objects(user_id):
+    time_query = epq.get_time_range_for_composite_object_creation(user_id)
+    ts = esta.TimeSeries.get_time_series(user_id)
+    confirmedTrips = esda.get_entries(esda.CONFIRMED_TRIP_KEY, user_id, time_query=time_query)
+    logging.info("Creating composite trips from %d confirmed trips" % len(confirmedTrips))
+    if len(confirmedTrips) == 0:
+        logging.debug("len(confirmedTrips) == 0, early return")
+        return None
+    for ct in confirmedTrips:
+        logging.info("End place type for trip is %s" %  type(ct.data.end_place))
+        end_place = esda.get_entry(esda.CONFIRMED_PLACE_KEY, ct["data"]["confirmed_place"])
+        composite_trip_dict = copy.copy(ct)
+        del composite_trip_dict["_id"]
+        composite_trip_dict["metadata"]["key"] = "analysis/composite_trip"
+        composite_trip_dict["data"]["confirmed_place"] = end_place
+
+        # later we will want to put section & modes in composite_trip as well
+
+        composite_trip_entry = ecwe.Entry(composite_trip_dict)
+        # save the entry
+        ts.insert(composite_trip_entry)
+    epq.mark_composite_object_creation_done(user_id, None)
+
 def create_confirmed_objects(user_id):
     time_query = epq.get_time_range_for_confirmed_object_creation(user_id)
     try:
-        last_expected_trip_done = create_confirmed_trips(user_id, time_query)
-        if last_expected_trip_done is None:
+        # we will use the same query for trips and places, but using 'exit_ts' instead 'end_ts'
+        time_query.timeType = "data.exit_ts"
+        last_expected_place_done = create_confirmed_places(user_id, time_query)
+        time_query.timeType = "data.end_ts"
+        last_expected_trip_done = create_confirmed_trips(user_id, time_query, last_expected_place_done["_id"])
+        if last_expected_trip_done is None or last_expected_place_done is None:
```

Comment:
This is correct from a high level perspective, but there are a couple of minor wrinkles that we (1) need to be careful about, (2) see how we have done before, and (3) test.

1. the last place (aka the place we are in right now) will not have an exit timestamp. So I would expect that it would not be returned by this query. That implies that we will not create a confirmed place for it. So if the user added any time use entries to the last place, I would expect that they would not visible to the user after the timeline has finished processing.
    a. Trip ends at 5:30, will not leave until the next morning
    b. Pipeline runs, we create a cleaned place with enter = 5:30
    c. user enters a time use entry
    d. entry is pushed to server, pipeline runs 8:30pm
    e. when we process the entry, it doesn't match any confirmed place because the cleaned place will not match the exit_ts query
    f. next morning we take the new trip
    g. next time pipeline runs, we will create a confirmed trip, and it will match the existing user input
The part that I am worried about is between steps d and f because the pipeline may think that it had finished processing the entry, so it will not be returned as an unprocessed entry and it will not be matched either, so it may seem to "disappear". We need to test this super carefully.
2. for the last trip, when we try to find the composite place at `end_place = esda.get_entry(esda.CONFIRMED_PLACE_KEY, ct["data"]["confirmed_place"])`, we won't find anything. So I think that we will not even display the final place?
3. should we mark the `done' timestamp as the trip or the place? If we use the trip, this code works except for the issue above because we will have generated confirmed trips until the final trip and the confirmed place will be generated after the next trip is done. But if we want to address the issue above, we may want to mark the `done` with the `enter_ts` of the last place instead. We should check on how we have done this for other stages that mix trips and places, namely `CLEAN_AND_RESAMPLE`

Reply:
So the last cleaned_place (where we are right now) will have an `enter_ts` but no `exit_ts`. I thought that we wouldn't want to create a `confirmed_place` that is incomplete in this way.

But if we do want to, can't we just query for cleaned places using  `enter_ts` instead? The last trip's `end_ts` is always the same as the last place's `enter_ts` right?

Reply:
I think that the two options that would work are:
1) do what we do for `CLEAN_AND_RESAMPLE`, which is that we keep the trip timestamp as the primary BUT when we handle a trip, we also handle it succeeding place in the step. So basically have a `create_composite_trip` and within that function, while processing trips also create the confirmed place that follows it.
2) or use `enter_ts` for places and keep it seperate

1) has the advantage that you can look at `CLEAN_AND_RESAMPLE` for a template and we know it works
2) has the advantage that it is cleaner conceptually since we are working on parts of the timeline independently and this does not appear to have cross component correlations as we do in the `CLEAN_AND_RESAMPLE`